Feat publish source

matnad opened the pull request:
### What I did

Added a method to publish source to etherscan for solidity contracts.
`DeployedContract.publish_source()`

Related issue: #411 

### How I did it
Flatten source code into one file (thanks Ben for starting on this). Make sure to include all comments, proper licensing and exactly one pragma statement. 
Will wait for contract to be recognized by etherscan after deploying, then submit request for verification and finally wait for the verification to be processed and report the result.

### How to verify it
`deployed = MyContract.deploy({'from': account}, publish_source=True)`
and also 
`MyContract.get_verification_info()`
![image](https://user-images.githubusercontent.com/18177452/103249528-fecd5880-496f-11eb-8ccf-5c2bd4437cf3.png)
-> https://ropsten.etherscan.io/address/0x6De9c6197bB8f61177C30946d90CAC2720272c5c#code


Tests are adapted for the changes to the solc versioning, but no new tests for this feature are written yet.

### Checklist

- [x] I have confirmed that my PR passes all linting checks
- [x] I have included test cases
- [x] I have updated the documentation
- [x] I have added an entry to the changelog

matnad commented:
@Macarse  @banteg feel free to test the branch and provide feedback!

iamdefinitelyahuman commented on brownie/network/contract.py:
```diff
@@ -530,6 +591,136 @@ def balance(self) -> Wei:
+            "sourceCode": flat["flattened_source"],
+            "codeformat": "solidity-single-file",
+            "contractname": flat["contract_name"],
+            "compilerversion": "".join(("v", flat["compiler_version"])),
```
What is the reason you didn't use an f-string here?

iamdefinitelyahuman commented on brownie/network/contract.py:
```diff
@@ -127,6 +128,66 @@ def decode_input(self, calldata: Union[str, bytes]) -> Tuple[str, Any]:
 
         return function_sig, input_args
 
+    def flatten(self) -> dict:
```
```suggestion
    def flatten(self) -> Dict:
```

iamdefinitelyahuman commented on brownie/network/contract.py:
```diff
@@ -127,6 +128,66 @@ def decode_input(self, calldata: Union[str, bytes]) -> Tuple[str, Any]:
+                f"pragma solidity {version_short};{flattened_source}\n\n"
+                f"// File: {file_name}\n\n{source}\n"
+            )
+            return {
```
I understand the importance of returning all this information for source verification, but with Vyper the function is only returning a source file.  In order to keep the return value consistent it might be better to simply raise when language == Vyper and say "you don't need to flatten this".  ?

matnad replied:
Agreed, as long as vyper verification is not implemented via API, we should raise.

iamdefinitelyahuman reviewed (approved):
:rocket:

matnad commented:
Update:
- requested changes
- Handle `pragma experimental ABIEncoderV2`
- Auto-fetch correct constructor arguments
- Temporary fix for interfaces being saved into `build/contracts/` (will discuss with @iamdefinitelyahuman for a better solution)

TODO:
- [ ]  Ensure correct order of files in the flattened file (base before derived is a hard requirement)
- [ ] Fix interface/contract issue

Macarse commented:
@iamdefinitelyahuman I have tested the last version: https://github.com/eth-brownie/brownie/pull/914/commits/763a5c827987a21632dac9d238d7d0afe4da4d96
 and I could verify a couple of contracts I couldn't verify in the past. I believe this is a good first version for contract verification 🚀 
 
 Great work @matnad h/t

matnad commented:
This turned out to be a monster of a feature! A lot of edge cases have been handled and I currently can't find a repo that doesn't verify (including all brownie mixes). **Happy to merge after a review.**

For repos that don't work please open an issue and I will look at it.

Macarse commented on brownie/network/contract.py:
```diff
@@ -127,6 +130,153 @@ def decode_input(self, calldata: Union[str, bytes]) -> Tuple[str, Any]:
+        elif language == "Solidity":
+            flattened_source = ""
+            has_abiencoder = False
+            abiencoder_re = r"(^|\r|\n|\r\n)\s*pragma experimental ABIEncoderV2;"
```
Should we move all this parsing logic into a separate class?

matnad replied:
That is Ben's call :P
The contract.py needs to be refactored anyways at some point... I don't think we should do that in this PR.

iamdefinitelyahuman replied:
This probably should only be available on the `ContractContainer`.  Otherwise someone will complain when their contract has a `get_verification_info` function and can't be used with brownie cuz of namespace collision.

Macarse commented on docs/api-network.rst:
```diff
@@ -297,7 +297,7 @@ Account Methods
         True
 
-.. py:classmethod:: Account.deploy(contract, *args, amount=None, gas_limit=None, gas_price=None, nonce=None, required_confs=1)
+.. py:classmethod:: Account.deploy(contract, *args, amount=None, gas_limit=None, gas_price=None, nonce=None, required_confs=1, allow_revert=False, silent=False, publish_source=False,)
```
I wonder if `silent` should be `True ` by default. I would like to avoid the print of `Verification complete. Result: Pass - Verified` for end users.

matnad replied:
I think some feedback is nice, especially for longer calls and for newer users. Advanced users will know how to disable it. 
@iamdefinitelyahuman you have an opinion on this?

iamdefinitelyahuman replied:
The `silent` kwarg is applying to both verification and deployment..  I don't think we'd want to silence the output on the actual deployment action, so unless we split it into 2 silent kwargs, I vote it's `False` by default.

matnad replied:
I think two args is overkill for this. Since verification takes around 30 secs+, I think it is nice to let the user know why it is stalling.

Macarse commented on tests/project/compiler/test_solidity.py:
```diff
@@ -58,11 +58,11 @@ def msolc(monkeypatch):
     compiler.set_solc_version("0.5.7")
-    assert solcx.get_solc_version().truncate() == compiler.solidity.get_version()
-    assert solcx.get_solc_version().truncate() == Version("0.5.7")
+    assert solcx.get_solc_version(True) == compiler.solidity.get_version()
```
no idea what the boolean is used of the `get_solc_version()` call but it looks like `True` is used in every instance in this PR. Shouldn't that be the default?

matnad replied:
It will be the default moving on, we had to make some compiler changes and to not break anything we added `True` to use the updated version. Once everything is changed and working, we can move to the new version entirely.

iamdefinitelyahuman replied:
For readability, suppose it could say `with_commit_hash=True`

iamdefinitelyahuman commented on brownie/network/contract.py:
```diff
@@ -127,6 +130,153 @@ def decode_input(self, calldata: Union[str, bytes]) -> Tuple[str, Any]:
+                # Create a duplicate entry for each import alias
+                part_name_short = part_name
+                if "/" in part_name:
+                    part_name_short = part_name.split("/")[-1]
```
We know for sure this is OK on windows, yes? Solidity always uses unix-style paths?

matnad replied:
Yes, this is taken from the build jsons, the path is always with /.

iamdefinitelyahuman commented on brownie/network/contract.py:
```diff
@@ -127,6 +130,153 @@ def decode_input(self, calldata: Union[str, bytes]) -> Tuple[str, Any]:
+                start = 0
+                for o in offsets:
+                    cleaned_section = remove_comments(src[start : o[0]])
+                    global_structs.update(re.findall(r"(\w*struct [^}]+})", cleaned_section))
```
It would be better to find these using the AST.  Regex is too prone to false-positives...  someone might have a commented out struct or enum.

iamdefinitelyahuman commented on brownie/network/contract.py:
```diff
@@ -530,6 +684,151 @@ def balance(self) -> Wei:
         balance = web3.eth.getBalance(self.address)
         return Wei(balance)
 
+    def publish_source(self, silent: bool = False) -> bool:
```
Same comment as above...  this is probably better as `ContractContainer.publish_source(contract)` where you hand it a `Contract` object.  Otherwise more collisions :scream:

iamdefinitelyahuman reviewed (commented):
Coming along nicely..

matnad commented:
Big update.
- Complete refactor of `get_verification_info()` to use py-solc-ast. No more regex (except one that is safe). This is much cleaner now!
- Moved the methods to `ContractContainer` and adjusted docs.
- Added a simple test to check if verification info breaks.
- Tested functionality with various repos (brownie mixes, private repos, testing repos).

iamdefinitelyahuman commented on brownie/network/contract.py:
```diff
@@ -239,6 +243,309 @@ def _add_from_tx(self, tx: TransactionReceiptType) -> None:
+            for n, src in [ns.values() for ns in nodes_source]:
+                pragmas = n.children(filters={"nodeType": "PragmaDirective"})
+                for pragma in pragmas:
+                    src_offsets = [int(x) for x in pragma.get("src").split(":")]
+                    pragma_statements.add(src[src_offsets[0] : src_offsets[0] + src_offsets[1]])
```
```suggestion
                    pragma_statements.add(src[slice(*pragma.offset)])
```

`solcast` node objects already expose the offset formatted in the way you require.

iamdefinitelyahuman replied:
I think this change can be applied in a few more places below

matnad replied:
Good catch, thanks!

iamdefinitelyahuman reviewed (approved):
Amazing stuff..  this clearly has been a huge amount of work.  The community loves you :heart:

misterblonde commented:
I still have the issue, even when I'm using `publish_source=True` and your new `get_verification_info()` afterwards. I only have one solidity script with one compiler version, so that shouldn't be the issue. Has this been continued elsewhere/

turkishfilms commented:
im getting a jsondecodererror when i use publish_source=True. I was able to manually verify(copy and paste) but wasnt able to automatically verify. only one .sol file
